openadapt/window/_windows.py

```diff
@@ -1,30 +1,155 @@
+        "height": meta["rectangle"].height(),
+        "meta": {**meta, "rectangle": rectangle_dict},
+        "data": data,
+        "window_id": meta["control_id"],
     }
+    try:
+        pickle.dumps(state)
+    except Exception as exc:
+        logger.warning(f"{exc=}")
+        state.pop("data")
     return state
 
 
-def get_element_at_position(x, y):
-    # TODO
-    return None
+def get_active_window_meta(active_window) -> dict:
+    """
+    Get the meta information of the active window.
+
+    Args:
+        active_window: The active window object.
+
+    Returns:
+        dict: A dictionary containing the meta information of the active window.
+    """
+    if not active_window:
+        logger.warning(f"{active_window=}")
+        return None
+    result = active_window.get_properties()
+    return result
+
+
+def get_active_element_state(x: int, y: int):
+    """
+    Get the state of the active element at the given coordinates.
+
+    Args:
+        x (int): The x-coordinate.
+        y (int): The y-coordinate.
+
+    Returns:
+        dict: A dictionary containing the properties of the active element.
+    """
+    active_window = get_active_window()
+    active_element = active_window.from_point(x, y)
+    properties = active_element.get_properties()
+    properties["rectangle"] = serialize_rect(properties["rectangle"])
+    return properties
+
+
+def get_active_window(depth=10, max_width=10, filename=None) -> Desktop:
+    """
+    Get the active window object.
+
+    Returns:
+        Desktop: The active window object.
+    """
+    app = pywinauto.application.Application(backend="uia").connect(active_only=True)
+    window = app.active()
+    return window
+
+
+def get_element_properties(element):
+    """
+    Recursively retrieves the properties of each element and its children.
+
+    Args:
+        element: An instance of a custom element class that has the `.get_properties()` and `.children()` methods.
+
+    Returns:
+        A nested dictionary containing the properties of each element and its children.
+        The dictionary includes a "children" key for each element, which holds the properties of its children.
+
+    Example:
+        element = Element()
+        properties = get_element_properties(element)
+        print(properties)
+        # Output: {'prop1': 'value1', 'prop2': 'value2', 'children': [{'prop1': 'child_value1', 'prop2': 'child_value2', 'children': []}]}
+    """
+
+    properties = element.get_properties()
+    children = element.children()
+
+    if children:
+        properties["children"] = [get_element_properties(child) for child in children]
+
+    # Serialize the "rectangle" key
+    properties["rectangle"] = serialize_rect(properties["rectangle"])
+
+    return properties
+
+
+def serialize_rect(rect):
+    serialized = {
+        "left": rect.left,
+        "top": rect.top,
+        "right": rect.right,
+        "bottom": rect.bottom,
+    }
+    return serialized
+
+
+def main():
```

Comment:
Maybe a docstring here would be useful, e.g. to indicate this is for testing?

Reply:
@abrichr I agree, done!